src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+    {
+        private readonly Symbol _enclosingSymbol;
+        private readonly DiagnosticBag _diagnostics;
+        private readonly BoundPatternSwitchStatement _switchStatement;
+        private readonly Conversions _conversions;
+
+        private CSharpSyntaxNode _syntax;
+
+        public DecisionTreeComputer(
+            Symbol enclosingSymbol,
+            DiagnosticBag diagnostics,
+            BoundPatternSwitchStatement switchStatement,
+            Conversions conversions)
+        {
+            this._enclosingSymbol = enclosingSymbol;
+            this._diagnostics = diagnostics;
+            this._switchStatement = switchStatement;
+            this._conversions = conversions;
+        }
+
+        internal DecisionTree ComputeDecisionTree()
+        {
+            var expressionIsNull = Binder.ExpressionIsNull(_switchStatement.Expression) != false;
+            var result = DecisionTree.Create(_switchStatement.Expression, _switchStatement.Expression.Type);
+            BoundPatternSwitchLabel defaultLabel = null;
+            foreach (var section in _switchStatement.SwitchSections)
+            {
+                foreach (var label in section.SwitchLabels)
+                {
+                    if (label.Syntax.Kind() == SyntaxKind.DefaultSwitchLabel)
+                    {
+                        if (defaultLabel != null)
+                        {
+                            // duplicate switch label will have been reported during initial binding.
+                        }
+                        else
+                        {
+                            defaultLabel = label;
+                        }
+                    }
+                    else
+                    {
+                        this._syntax = label.Syntax;
+                        var subsumedErrorCode = CheckSubsumed(label.Pattern, result, expressionIsNull);
+                        if (subsumedErrorCode != 0 && subsumedErrorCode != ErrorCode.ERR_NoImplicitConvCast)
+                        {
+                            if (!label.HasErrors)
+                            {
+                                _diagnostics.Add(subsumedErrorCode, label.Pattern.Syntax.Location);
+                            }
+                        }
+                        else
+                        {
+                            AddToDecisionTree(result, label.Pattern, label.Guard, label);
+                        }
+                    }
+                }
+            }
+
+            if (defaultLabel != null)
+            {
+                Add(result, (e, t) => new DecisionTree.Guarded(_switchStatement.Expression, _switchStatement.Expression.Type, default(ImmutableArray<KeyValuePair<BoundExpression, LocalSymbol>>), null, defaultLabel));
+            }
+
+            return result;
+        }
+
+        private void AddToDecisionTree(DecisionTree decisionTree, BoundPattern pattern, BoundExpression guard, BoundPatternSwitchLabel label)
```

Comment:
Sure. These are likely to be needed when the code is generalized for the match statement, but I can reintroduce the parameters at that time.

---
In reply to: [64672084](https://github.com/dotnet/roslyn/pull/11485#discussion_r64672084) [](ancestors = 64672084)